src/backend/distributed/executor/multi_utility.c

```diff
@@ -1623,211 +1651,421 @@ ErrorIfUnsupportedAlterTableStmt(AlterTableStmt *alterTableStatement)
+			referencedTableColocationId = TableColocationId(referencedTableId);
+			if (colocationId == INVALID_COLOCATION_ID ||
+				colocationId != referencedTableColocationId)
 			{
 				ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
-								errmsg("alter table command is currently unsupported"),
-								errdetail("Only ADD|DROP COLUMN, SET|DROP NOT NULL,"
-										  " SET|DROP DEFAULT, ADD|DROP CONSTRAINT FOREIGN"
-										  " KEY and TYPE subcommands are supported.")));
+								errmsg("cannot create foreign key constraint"),
+								errdetail("Foreign key constraint can only be created"
+										  " on co-located tables.")));
+			}
+
+			/*
+			 * Partition column must exist in both referencing and referenced side of the
+			 * foreign key constraint. They also must be in same ordinal.
+			 */
+			referencedTablePartitionColumn = PartitionKey(referencedTableId);
+		}
+		else
+		{
+			/*
+			 * Partition column must exist in both referencing and referenced side of the
+			 * foreign key constraint. They also must be in same ordinal.
+			 */
+			referencedTablePartitionColumn = distributionColumn;
+		}
+
+		/*
+		 * Column attributes are not available in Form_pg_constraint, therefore we need
+		 * to find them in the system catalog. After finding them, we iterate over column
+		 * attributes together because partition column must be at the same place in both
+		 * referencing and referenced side of the foreign key constraint
+		 */
+		referencingColumnsDatum = SysCacheGetAttr(CONSTROID, heapTuple,
+												  Anum_pg_constraint_conkey, &isNull);
+		referencedColumnsDatum = SysCacheGetAttr(CONSTROID, heapTuple,
+												 Anum_pg_constraint_confkey, &isNull);
+
+		deconstruct_array(DatumGetArrayTypeP(referencingColumnsDatum), INT2OID, 2, true,
+						  's', &referencingColumnArray, NULL, &referencingColumnCount);
+		deconstruct_array(DatumGetArrayTypeP(referencedColumnsDatum), INT2OID, 2, true,
+						  's', &referencedColumnArray, NULL, &referencedColumnCount);
+
+		Assert(referencingColumnCount == referencedColumnCount);
+
+		for (attrIdx = 0; attrIdx < referencingColumnCount; ++attrIdx)
+		{
+			AttrNumber referencingAttrNo = DatumGetInt16(referencingColumnArray[attrIdx]);
+			AttrNumber referencedAttrNo = DatumGetInt16(referencedColumnArray[attrIdx]);
+
+			if (distributionColumn->varattno == referencingAttrNo &&
+				referencedTablePartitionColumn->varattno == referencedAttrNo)
+			{
+				foreignConstraintOnPartitionColumn = true;
 			}
 		}
+
+		if (!foreignConstraintOnPartitionColumn)
+		{
+			ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("cannot create foreign key constraint"),
+							errdetail("Partition column must exist both "
+									  "referencing and referenced side of the "
+									  "foreign constraint statement and it must "
+									  "be in the same ordinal in both sides.")));
+		}
+
+		multiShardedReferencingExist = IsDistributedTable(referencingTableId) &&
```

Comment:
This part is confusing.  I assume `multiShardedReferencingExist` means `IsDistributedTable(referencingTableId)` and `multiShardedReferencingBeingCreated ` means `!IsDistributedTable(referencingTableId)`. `!SingleReplicatedTable(referencingTableId)` and `ShardReplicationFactor > 1` are only exist to use in if below.

What are you trying to do here? I presume you want to distinguish between `CREATE TABLE` and `ALTER TABLE` and check `shard_count > 1` according to that. If we rewrite this in this form, will it be correct?

```c
if (IsDistributedTable(referencingTableId))
  {
    \\ ALTER TABLE...
    if (!SingleReplicatedTable(referencingTableId))
    {
        singleReplicatedTable = false;
     }
}
else
{
  \\ CREATE TABLE
  if (ShardReplicationFactor > 1)
  {
       singleReplicatedTable = false;
  }
}

if(!singleReplicatedTable)
{
 \\ error out
}
```

You may also want to use `selfReferencingTable` in checks above. What is our policy for self-referencing tables? Do we support constraints on them all the time?

Does this make sense? What do you think?